wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,298 @@
+            This directory is captured in |document_root|. It also suppresses
+            logging."""
+
+            def translate_path(self, path):
+                path = SimpleHTTPRequestHandler.translate_path(self, path)
+                requested_uri = os.path.relpath(path, os.getcwd())
+                return os.path.join(document_root, requested_uri)
+
+            # Disable the logging.
+            # pylint: disable=redefined-builtin
+            def log_message(self, format, *args):
+                pass
+
+        class _ArchiveServerThread(threading.Thread):
+            """Thread for running the HTTPServer"""
+
+            def __init__(self, httpd):
+                self._httpd = httpd
+                threading.Thread.__init__(self)
+
+            def run(self):
+                self._httpd.serve_forever()
+
+        # Create the server, and find out the port we've been assigned...
+        self._httpd = HTTPServer(("", 0), _DifferentDirectoryHTTPRequestHandler)
+        _, self._port = self._httpd.server_address
+
+        self._thread = _ArchiveServerThread(self._httpd)
+        self._thread.start()
+
+        adb_command(target.adb_name, "reverse tcp:{0} tcp:{0}".format(self._port))
+
+    def stop(self, target):
+        adb_command(target.adb_name, "reverse --remove tcp:{}".format(self._port))
+
+        self._httpd.shutdown()
+        self._thread.join()
+
+    def get_port(self):
+        return self._port
+
+
+class Speedometer(Workload):
+
+    name = "speedometer"
+    description = """
+    A workload to execute the speedometer 2.0 web based benchmark. Requires device to be rooted.
 
     Test description:
-    1. Open chrome
-    2. Navigate to the speedometer website - http://browserbench.org/Speedometer/
-    3. Execute the benchmark
+    1. Host a local copy of the Speedometer website, and make it visible to the device via ADB.
+    2. Open chrome via an intent to access the local copy.
+    3. Execute the benchmark - the copy has been modified to trigger the start of the benchmark.
 
-    known working chrome version 80.0.3987.149
-    '''
+    known working chrome version 83.0.4103.106
+    """
+    supported_platforms = ["android"]
+
+    package_names = ["org.chromium.chrome", "com.android.chrome"]
+    regex = re.compile('text="(\d+.\d+)" resource-id="result-number"')
 
     parameters = [
-        Parameter('speedometer_version', allowed_values=versions, kind=str, default='2.0',
-                  description='''
-                  The speedometer version to be used.
-                  ''')
+        Parameter(
+            "chrome_package",
+            allowed_values=package_names,
+            kind=str,
+            default="com.android.chrome",
+            description="""
+                  The app package for the browser that will be launched.
+                  """,
+        ),
+        Parameter(
+            "render_process_search_term",
```

Comment:
This seems to be a cop-out, and the whole process seems very brittle...

Since you are modifying the Speedometer code anyway, could you not also get it to signal its own completion, e.g. by creating a file, or opening an alert dialog that you can then look for with UIAutomatior?